### ISSUE TITLE
table.cell not_last_on_page argument

### ISSUE THREAD

**Issue by bugsbugsbux** [typst/typst#4841]
### Description

It would be awesome to have a table.cell argument "not_last_on_page:true" to prevent a cell from being in the last row on a page (when the table is split).

I wonder if the option would better just be called "not_last" since the same behaviour should apply when the table splits in a multi-column layout?

### Use Case

This would be useful in tables which have "subheadings":

```typ
#table(columns: 2,
    table.header([*First Name*], [*Last Name*]),
    table.cell(colspan: 2, fill: silver, align: center, [Employees]),
    [Alice], [Abrams],
    [Bob], [Brown],
    table.cell(colspan: 2, fill: silver, align: center, [Customers]),
    [Charlie], [Chaplin],
    [David], [Davidson],
)
```

**Comment by bluebear94** [comment]
This option could alternatively be called “keep with next row” (analogously to [“keep with next paragraph”](https://practicaltypography.com/keep-with-next-paragraph.html) in paragraph layout).

How would you specify this option for a row with multiple cells? Perhaps turn it on if any cell in that row has the option?

**Comment by laurmaedje** [comment]
As a workaround, what might work is adding a zero-width column and then adding rowspanned unbreakable cells at the appropriate positions. (I haven't tested this.)

**Comment by PgBiel** [comment]
> As a workaround, what might work is adding a zero-width column and then adding rowspanned unbreakable cells at the appropriate positions. (I haven't tested this.)

yeah this would probably work (would become an internal row group).

More generally, we could expose a row group table primitive, similarly to `table.header`.

**Comment by bugsbugsbux** [comment]
> How would you specify this option for a row with multiple cells? Perhaps turn it on if any cell in that row has the option?

Yes, exactly. I believe this is also how `table.cell`'s `breakable=true` works...

**Comment by Marco-Kaulea** [comment]
We have a related, but I'm not sure the same issue. We want to prevent single rows (orphaned/widowed) both at the start and end of pages. The workaround with the 0-width column and unbreakable rowspan cell works, but is quite cumbersome, because the table isn't hardcoded. Also, these are basically subtables embedded in a cell of a larger table, I don't understand enough of typst's internals to know whether this makes a solution more difficult.
If this is a different problem, I can create a separate issue.

**Comment by PgBiel** [comment]
@Marco-Kaulea I feel like #6168 will solve your problem, and also that of the original poster.

In particular, you will be able to have multiple headers and optionally repeat them across pages. They all have orphan prevention.

For widow prevention, this will require subfooters which are still a work in progress.

**Comment by Marco-Kaulea** [comment]
@PgBiel Thanks for the quick response. I'm not entirely sure it will, we don't have an issue with headers, we (or more specifically the person deciding how the document should look) want to avoid normal table rows standing by themselves.

**Comment by mewmew** [comment]
The "subheader" use case mentioned in https://github.com/typst/typst/issues/4841#issue-2487172296  was resolved by #6168.

## Example (before using subheader feature)

Example before using subheader feature of #6168 (using 3 data rows per subheader, just to make the example clearer).

```typst
#set page(height: 3.5cm, width: 5.5cm)

#table(columns: 2,
	table.header([*First Name*], [*Last Name*]),
	table.cell(colspan: 2, fill: silver, align: center, [Employees]),
	[Alice], [Abrams],
	[Bob], [Brown],
	[Charlie], [Chaplin],
	table.cell(colspan: 2, fill: silver, align: center, [Customers]),
	[David], [Davidson],
	[Eric], [Ericson],
	[Fredric], [Fredricson],
)
```

<img width="482" height="914" alt="Image" src="https://github.com/user-attachments/assets/a7e040ab-4fc3-43ff-8266-3dae544dc00c" />

## Example, resolved (after using subheader feature)

The updated example (using the subheader feature of #6168) would be as follows:

```typst
#set page(height: 3.5cm, width: 5.5cm)

#table(columns: 2,
	table.header([*First Name*], [*Last Name*]),
	table.header(
		level: 2,
		table.cell(colspan: 2, fill: silver, align: center, [Employees]),
	),
	[Alice], [Abrams],
	[Bob], [Brown],
	[Charlie], [Chaplin],
	table.header(
		level: 2,
		table.cell(colspan: 2, fill: silver, align: center, [Customers]),
	),
	[David], [Davidson],
	[Eric], [Ericson],
	[Fredric], [Fredricson],
)
```

<img width="456" height="1186" alt="Image" src="https://github.com/user-attachments/assets/315380d6-2d05-4d59-8d71-697e70a01a09" />

It seems for most common use cases, #6168 should be enough to close this issue.

**Edit:** I was too quick to post the comment, and didn't realize it was already answered in https://github.com/typst/typst/issues/4841#issuecomment-2933740185

**Comment by bugsbugsbux** [comment]
I don't think it's a good idea to misuse elements for something generally useful like this. It would mean, that anytime I style my table headers with a general rule, it affects table rows which are not really headers but just use this element to make them not the last row on a page. I want to be able to write my table and say "oh, reading this row, one really shouldn't miss the next one due to it being moved to the next page..."

**Comment by danvim** [comment]
Ideally, I'd like to have the ability to set some row-group to avoid breaking inside, but it should still break across pages if it absolutely need to.

Adding a 0-width column also affects accessibility by having extra structural elements.
